<a href="https://colab.research.google.com/github/christophermalone/DSCI325/blob/main/Module3_Part2R.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Module 3 | Part 2 R: Data Verb - FILTER()

<table width='100%' ><tr><td bgcolor='green'></td></tr></table>

### Example 3.2.R
For this notebook, we will consider airline data from the Bureau of Transportation.  Using the form provided on their website, one is able to obtain a variety of information around flight delays.
 

The following 17 fields will be considered here:

*   Day Information: DAY_OF_MONTH, DAY_OF_WEEK
*   Origin Information: ORIGIN, ORIGIN_STATE
*   Destination Information: DEST, DEST_STATE
*   Departure Information: DEP_TIME, DEP_DELAY, DEP_DELAY15, 
*   Arrival Information: ARR_TIME, ARR_DELAY, ARR_DEL15, 
*   Reason for Delay: CARRIER_DELAY, WEATHER_DELAY, SECURITY_DELAY, LATE_AIRCRAFT_DELAY


<br>Data Source:  https://www.transtats.bts.gov/DL_SelectFields.asp?gnoyr_VQ=FGJ&QO_fu146_anzr=b0-gvzr

<table width='100%' ><tr><td bgcolor='green'></td></tr></table>


To begin, the <strong>tidyverse</strong> package will be used throughout this notebook; thus, load this package.

In [ ]:
library(tidyverse)

── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──

✔ ggplot2 3.3.3     ✔ purrr   0.3.4
✔ tibble  3.0.6     ✔ stringr 1.4.0
✔ tidyr   1.1.2     ✔ forcats 0.5.1
✔ readr   1.4.0     

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()



Next, read the Ontime_Reporting.csv file into R.

In [ ]:
Ontime<-read_csv('/content/sample_data/Ontime_Reporting.csv')

Next, use dim() to get the number of rows and columns and the head() function to see the first few rows of this dataframe.

In [ ]:
#Getting the dimension of the dataframe and taking a look at a few rows.
 
 cat("The dimension of the Ontime dataframe is: \n")
   dim(Ontime)
 
 cat("\n")
 
 cat("A printout of the Ontime dataframe:\n")
   head(Ontime)
   

The dimension of the Ontime dataframe is: 


[1] 371357     17


A printout of the Ontime dataframe:


DAY_OF_MONTH DAY_OF_WEEK ORIGIN ORIGIN_STATE DEST DEST_STATE DEP_TIME
1 1            2           CLT    NC           MCO  FL         1252    
2 1            2           MCO    FL           CLT  NC         1525    
3 1            2           DFW    TX           MCO  FL          840    
4 1            2           MCO    FL           DFW  TX         1328    
5 1            2           EWR    NJ           DFW  TX          604    
6 1            2           MEM    TN           CLT  NC          952    
  DEP_DELAY DEP_DEL15 ARR_TIME ARR_DELAY ARR_DEL15 CARRIER_DELAY WEATHER_DELAY
1  -7       0         1421     -19       0         NA            NA           
2 -11       0         1701     -20       0         NA            NA           
3  -5       0         1200     -13       0         NA            NA           
4  -5       0         1530      -5       0         NA            NA           
5  -6       0          835     -47       0         NA            NA           
6  19       1         1225       8       0         NA            NA           
  NAS_DELAY SECURITY_DELAY LATE_AIRCRAFT_DELAY
1 NA        NA             NA                 
2 NA        NA             NA                 
3 NA        NA             NA                 
4 NA        NA             NA                 
5 NA        NA             NA                 
6 NA        NA             NA

# Filter Actions on Strings

To begin, suppose the goal is to obtain only flights whose ORIGIN airport was Rochester, MN.  The airport code for Rochester, MN is <strong>RST</strong>.

<p align='center'><img src="https://drive.google.com/uc?export=view&id=1cZ49YcqPChfiBZP0Hq7Ahzxn4Sg3ktG-" width='25%' height='25%'></p>

In [ ]:
#Getting only flights that left RST
(
  Ontime
  %>% filter(ORIGIN == "RST")

) -> RST


Looking at the dimension and a few rows of the RST dataframe.

In [75]:
#Getting the dimension of the dataframe and taking a look at a few rows.
 
 cat("The dimension of the RST dataframe is: \n")
   dim(RST)
 
 cat("\n")
 
 cat("A printout of the RST dataframe:\n")
   head(RST)
   

The dimension of the RST dataframe is: 


[1] 238  17


A printout of the RST dataframe:


DAY_OF_MONTH DAY_OF_WEEK ORIGIN ORIGIN_STATE DEST DEST_STATE DEP_TIME
1 1            2           RST    MN           DEN  CO         1104    
2 1            2           RST    MN           ORD  IL         1715    
3 1            2           RST    MN           ORD  IL          620    
4 1            2           RST    MN           MSP  MN          604    
5 1            2           RST    MN           MSP  MN         1155    
6 1            2           RST    MN           MSP  MN         1754    
  DEP_DELAY DEP_DEL15 ARR_TIME ARR_DELAY ARR_DEL15 CARRIER_DELAY WEATHER_DELAY
1 -11       0         1213     -33       0         NA            NA           
2 -10       0         1828     -22       0         NA            NA           
3 -10       0          747     -10       0         NA            NA           
4 -11       0          648     -33       0         NA            NA           
5  -5       0         1239     -15       0         NA            NA           
6  -5       0         1844     -19       0         NA            NA           
  NAS_DELAY SECURITY_DELAY LATE_AIRCRAFT_DELAY
1 NA        NA             NA                 
2 NA        NA             NA                 
3 NA        NA             NA                 
4 NA        NA             NA                 
5 NA        NA             NA                 
6 NA        NA             NA

Next, let us collect the rows where the ORIGIN airport is RST and the destination airport is MSP, i.e. Minneapolis, MN.

<p align='center'><img src="https://drive.google.com/uc?export=view&id=1c142gdrEPwS_j09pqK1x8kpQjYDYSs47" width='25%' height='25%'></p>

In [76]:
#Getting desired rows
(
  Ontime
  %>% filter(ORIGIN == "RST", DEST == "MSP")

) -> RST_to_MSP


Looking at the dimension and a few rows of the RST_to_MSP dataframe.

In [80]:
#Getting the dimension of the dataframe and taking a look at a few rows.
 
 cat("The dimension of the RST_to_MSP dataframe is: \n")
   dim(RST_to_MSP)
 
 cat("\n")
 
 cat("A printout of the RST_to_MSP dataframe:\n")
   head(RST_to_MSP)


The dimension of the RST_to_MSP dataframe is: 


[1] 118  17


A printout of the RST_to_MSP dataframe:


DAY_OF_MONTH DAY_OF_WEEK ORIGIN ORIGIN_STATE DEST DEST_STATE DEP_TIME
1 1            2           RST    MN           MSP  MN          604    
2 1            2           RST    MN           MSP  MN         1155    
3 1            2           RST    MN           MSP  MN         1754    
4 1            2           RST    MN           MSP  MN         1540    
5 2            3           RST    MN           MSP  MN          606    
6 2            3           RST    MN           MSP  MN         1149    
  DEP_DELAY DEP_DEL15 ARR_TIME ARR_DELAY ARR_DEL15 CARRIER_DELAY WEATHER_DELAY
1 -11       0          648     -33       0         NA            NA           
2  -5       0         1239     -15       0         NA            NA           
3  -5       0         1844     -19       0         NA            NA           
4  -5       0         1628     -13       0         NA            NA           
5  -9       0          658     -23       0         NA            NA           
6 -10       0         1224     -29       0         NA            NA           
  NAS_DELAY SECURITY_DELAY LATE_AIRCRAFT_DELAY
1 NA        NA             NA                 
2 NA        NA             NA                 
3 NA        NA             NA                 
4 NA        NA             NA                 
5 NA        NA             NA                 
6 NA        NA             NA

Next, let us collect the rows where the ORIGIN airport is RST or the ORIGIN airport is DLH, i.e. Duluth, MN.

<p align='center'><img src="https://drive.google.com/uc?export=view&id=1zQzTj9PJkooEfkkaioffNW1O49Cocy7Y" width='25%' height='25%'></p>




In [78]:
#Getting desired rows
(
  Ontime
  %>% filter(ORIGIN == "RST" | ORIGIN == "DLH")

) -> RST_or_DLH

Looking at the dimension and a few rows of the RST_to_MSP dataframe.

In [79]:
#Getting the dimension of the dataframe and taking a look at a few rows.
 
 cat("The dimension of the RST_or_DLH dataframe is: \n")
   dim(RST_or_DLH)
 
 cat("\n")
 
 cat("A printout of the RST_or_DLH dataframe:\n")
   head(RST_or_DLH)

The dimension of the RST dataframe is: 


[1] 359  17


A printout of the RST dataframe:


DAY_OF_MONTH DAY_OF_WEEK ORIGIN ORIGIN_STATE DEST DEST_STATE DEP_TIME
1 1            2           RST    MN           DEN  CO         1104    
2 1            2           RST    MN           ORD  IL         1715    
3 1            2           RST    MN           ORD  IL          620    
4 1            2           RST    MN           MSP  MN          604    
5 1            2           DLH    MN           MSP  MN         1602    
6 1            2           DLH    MN           MSP  MN          554    
  DEP_DELAY DEP_DEL15 ARR_TIME ARR_DELAY ARR_DEL15 CARRIER_DELAY WEATHER_DELAY
1 -11       0         1213     -33       0         NA            NA           
2 -10       0         1828     -22       0         NA            NA           
3 -10       0          747     -10       0         NA            NA           
4 -11       0          648     -33       0         NA            NA           
5  -6       0         1651     -27       0         NA            NA           
6  -6       0          644     -27       0         NA            NA           
  NAS_DELAY SECURITY_DELAY LATE_AIRCRAFT_DELAY
1 NA        NA             NA                 
2 NA        NA             NA                 
3 NA        NA             NA                 
4 NA        NA             NA                 
5 NA        NA             NA                 
6 NA        NA             NA

Next, let us collect the rows where the ORIGIN airport is RST or the ORIGIN airport is DLH, i.e. Duluth, MN and the destination is MSP.


Requested Rows: (ORIGIN is RST OR DLH) AND (DEST is MSP)

In [81]:
#Getting desired rows
(
  Ontime
  %>% filter(ORIGIN == "RST" | ORIGIN == "DLH")
  %>% filter(DEST == "MSP")
) -> RST_or_DLH_to_MSP


cat("The dimension of the RST_or_DLH_to_MSP dataframe is: \n")
   dim(RST_or_DLH_to_MSP)
 cat("\n")
 cat("A printout of the RST_or_DLH_to_MSP dataframe:\n")
   head(RST_or_DLH_to_MSP,20)


The dimension of the RST_or_DLH_to_MSP dataframe is: 


[1] 239  17


A printout of the RST_or_DLH_to_MSP dataframe:


DAY_OF_MONTH DAY_OF_WEEK ORIGIN ORIGIN_STATE DEST DEST_STATE DEP_TIME
1  1            2           RST    MN           MSP  MN          604    
2  1            2           DLH    MN           MSP  MN         1602    
3  1            2           DLH    MN           MSP  MN          554    
4  1            2           DLH    MN           MSP  MN         1331    
5  1            2           RST    MN           MSP  MN         1155    
6  1            2           DLH    MN           MSP  MN         1750    
7  1            2           DLH    MN           MSP  MN         1046    
8  1            2           RST    MN           MSP  MN         1754    
9  1            2           RST    MN           MSP  MN         1540    
10 2            3           RST    MN           MSP  MN          606    
11 2            3           DLH    MN           MSP  MN         1223    
12 2            3           DLH    MN           MSP  MN         1525    
13 2            3           DLH    MN           MSP  MN          552    
14 2            3           RST    MN           MSP  MN         1149    
15 2            3           DLH    MN           MSP  MN         1750    
16 2            3           RST    MN           MSP  MN         1754    
17 2            3           RST    MN           MSP  MN         1540    
18 3            4           RST    MN           MSP  MN          610    
19 3            4           DLH    MN           MSP  MN         1219    
20 3            4           DLH    MN           MSP  MN         1526    
   DEP_DELAY DEP_DEL15 ARR_TIME ARR_DELAY ARR_DEL15 CARRIER_DELAY WEATHER_DELAY
1  -11       0          648     -33       0         NA            NA           
2   -6       0         1651     -27       0         NA            NA           
3   -6       0          644     -27       0         NA            NA           
4   31       1         1416      12       0         NA            NA           
5   -5       0         1239     -15       0         NA            NA           
6  -10       0         1853     -18       0         NA            NA           
7   -6       0         1128     -32       0         NA            NA           
8   -5       0         1844     -19       0         NA            NA           
9   -5       0         1628     -13       0         NA            NA           
10  -9       0          658     -23       0         NA            NA           
11  -7       0         1307     -32       0         NA            NA           
12 -10       0         1613     -32       0         NA            NA           
13  -8       0          704     -10       0         NA            NA           
14 -10       0         1224     -29       0         NA            NA           
15 -10       0         1849     -21       0         NA            NA           
16  -5       0         1837     -26       0         NA            NA           
17  -5       0         1623     -17       0         NA            NA           
18  -5       0          706     -14       0         NA            NA           
19 -11       0         1310     -29       0         NA            NA           
20  -9       0         1612     -33       0         NA            NA           
   NAS_DELAY SECURITY_DELAY LATE_AIRCRAFT_DELAY
1  NA        NA             NA                 
2  NA        NA             NA                 
3  NA        NA             NA                 
4  NA        NA             NA                 
5  NA        NA             NA                 
6  NA        NA             NA                 
7  NA        NA             NA                 
8  NA        NA             NA                 
9  NA        NA             NA                 
10 NA        NA             NA                 
11 NA        NA             NA                 
12 NA        NA             NA                 
13 NA        NA             NA                 
14 NA        NA             NA                 
15 NA        NA             NA                 
16 NA        NA       

Next, let us collect the rows where the ORIGIN airport is anywhere in MN and destination is MSP.



In [82]:
#Getting desired rows
(
  Ontime
  %>% filter(ORIGIN_STATE == "MN")
  %>% filter(DEST == "MSP")
) -> MN_to_MSP

cat("The dimension of the MN_to_MSP dataframe is: \n")
   dim(MN_to_MSP)
cat("\n")
cat("A printout of the MN_to_MSP dataframe:\n")
   head(MN_to_MSP,20)

The dimension of the MN_to_MSP dataframe is: 


[1] 460  17


A printout of the MN_to_MSP dataframe:


DAY_OF_MONTH DAY_OF_WEEK ORIGIN ORIGIN_STATE DEST DEST_STATE DEP_TIME
1  1            2           INL    MN           MSP  MN         1540    
2  1            2           HIB    MN           MSP  MN          615    
3  1            2           BJI    MN           MSP  MN          603    
4  1            2           BJI    MN           MSP  MN         1302    
5  1            2           BRD    MN           MSP  MN          613    
6  1            2           RST    MN           MSP  MN          604    
7  1            2           DLH    MN           MSP  MN         1602    
8  1            2           DLH    MN           MSP  MN          554    
9  1            2           DLH    MN           MSP  MN         1331    
10 1            2           RST    MN           MSP  MN         1155    
11 1            2           DLH    MN           MSP  MN         1750    
12 1            2           DLH    MN           MSP  MN         1046    
13 1            2           RST    MN           MSP  MN         1754    
14 1            2           RST    MN           MSP  MN         1540    
15 2            3           INL    MN           MSP  MN         1729    
16 2            3           INL    MN           MSP  MN         1048    
17 2            3           BRD    MN           MSP  MN         1753    
18 2            3           HIB    MN           MSP  MN          606    
19 2            3           BJI    MN           MSP  MN          615    
20 2            3           BJI    MN           MSP  MN         1305    
   DEP_DELAY DEP_DEL15 ARR_TIME ARR_DELAY ARR_DEL15 CARRIER_DELAY WEATHER_DELAY
1   -5       0         1632     -27       0         NA            NA           
2    0       0          716      -4       0         NA            NA           
3  -12       0          702     -17       0         NA            NA           
4   -8       0         1348     -24       0         NA            NA           
5   -2       0          649     -20       0         NA            NA           
6  -11       0          648     -33       0         NA            NA           
7   -6       0         1651     -27       0         NA            NA           
8   -6       0          644     -27       0         NA            NA           
9   31       1         1416      12       0         NA            NA           
10  -5       0         1239     -15       0         NA            NA           
11 -10       0         1853     -18       0         NA            NA           
12  -6       0         1128     -32       0         NA            NA           
13  -5       0         1844     -19       0         NA            NA           
14  -5       0         1628     -13       0         NA            NA           
15 -21       0         1846     -20       0         NA            NA           
16  -7       0         1200      -7       0         NA            NA           
17  -2       0         1832     -18       0         NA            NA           
18  -9       0          705     -15       0         NA            NA           
19   0       0          712      -7       0         NA            NA           
20  -5       0         1353     -19       0         NA            NA           
   NAS_DELAY SECURITY_DELAY LATE_AIRCRAFT_DELAY
1  NA        NA             NA                 
2  NA        NA             NA                 
3  NA        NA             NA                 
4  NA        NA             NA                 
5  NA        NA             NA                 
6  NA        NA             NA                 
7  NA        NA             NA                 
8  NA        NA             NA                 
9  NA        NA             NA                 
10 NA        NA             NA                 
11 NA        NA             NA                 
12 NA        NA             NA                 
13 NA        NA             NA                 
14 NA        NA             NA                 
15 NA        NA             NA                 
16 NA        NA       

Question:  What happens to departures from MSP and arrivals to MSP?

In [ ]:
#Getting desired rows
(
  Ontime
  %>% filter(ORIGIN_STATE == "MN")
) -> MN

dim(MN)

[1] 9371   17

DAY_OF_MONTH DAY_OF_WEEK ORIGIN ORIGIN_STATE DEST DEST_STATE DEP_TIME
1 1            2           MSP    MN           DFW  TX         1634    
2 1            2           MSP    MN           CLT  NC         1342    
3 1            2           MSP    MN           DFW  TX          616    
4 1            2           MSP    MN           CLT  NC          635    
5 1            2           MSP    MN           PHX  AZ         1105    
6 1            2           MSP    MN           PHX  AZ         1531    
  DEP_DELAY DEP_DEL15 ARR_TIME ARR_DELAY ARR_DEL15 CARRIER_DELAY WEATHER_DELAY
1 31        1         1857      11       0         NA            NA           
2 -4        0         1711      -8       0         NA            NA           
3  1        0          856     -20       0         NA            NA           
4 -6        0         1032       2       0         NA            NA           
5 -5        0         1324     -16       0         NA            NA           
6 -9        0         1745     -27       0         NA            NA           
  NAS_DELAY SECURITY_DELAY LATE_AIRCRAFT_DELAY
1 NA        NA             NA                 
2 NA        NA             NA                 
3 NA        NA             NA                 
4 NA        NA             NA                 
5 NA        NA             NA                 
6 NA        NA             NA

Next, let us consider the various airports across MN.  


*   How many are there?
*   Where are they located?



In [83]:
#Airports across MN
(
  Ontime
  %>% filter(ORIGIN_STATE == "MN")
  %>% group_by(ORIGIN)
  %>% summarise(Count=n())
 ) 

ORIGIN Count
1 BJI      62 
2 BRD      53 
3 DLH     121 
4 HIB      53 
5 INL      53 
6 MSP    8778 
7 RST     238 
8 STC      13

Another way to identify the various airports is to use the <strong>distinct()</strong> function.

In [84]:
#Getting only RST
(
  Ontime
  %>% filter(ORIGIN_STATE == "MN")
  %>% distinct(ORIGIN)
 ) 

ORIGIN
1 MSP   
2 RST   
3 INL   
4 HIB   
5 BJI   
6 BRD   
7 DLH   
8 STC

Next, let us collect the rows where the ORIGIN airport is in MN and purposely exclude MSP.  This will be done here with a sequence of <strong>OR</strong> statements.




<p align='center'><img src="https://drive.google.com/uc?export=view&id=1hMSS_XOgk5izrSa1PNvKSiWkM62c9xMn" width='25%' height='25%'></p>

In [90]:
#Getting the desired rows
(
  Ontime
  %>% filter(ORIGIN == "BJI" | ORIGIN == "BRD" | ORIGIN == "DLH" | ORIGIN == "HIB" | ORIGIN == "INL" | ORIGIN == "RST" | ORIGIN == "STC"  )
 ) -> MN

cat("The dimension of the MN dataframe is: \n")
   dim(MN)


The dimension of the MN dataframe is: 


[1] 593  17

Next, let us collect the rows where the ORIGIN airport is in MN and purposely exclude MSP.  This code will use the the <strong>IN()</strong> option instead of a sequence of OR statements.

In [91]:
#Getting the desired rows
(
  Ontime
  %>% filter(ORIGIN %in% c("BJI", "BRD", "DLH", "HIB","INL", "RST", "STC") )
 ) -> MN

cat("The dimension of the MN dataframe is: \n")
   dim(MN)
cat("\n")



The dimension of the MN dataframe is: 


[1] 593  17

One more iteration for ORIGIN airport is in MN and purposely exclude MSP.  This code will use the the <strong>!=()</strong>, i.e. not equal, option instead of a sequence of OR statements or IN() option.

In [92]:
#Getting only RST
(
  Ontime
  %>% filter(ORIGIN_STATE == "MN")
  %>% filter(ORIGIN != "MSP")
 ) -> MN

cat("The dimension of the MN dataframe is: \n")
   dim(MN)
cat("\n")

The dimension of the MN dataframe is: 


[1] 593  17

# Filter Action with NA

The following code is used to <strong>drop all </strong> missing from a dataframe.

In [94]:
#Getting desired rows
(
  Ontime
  %>% filter(ORIGIN == "RST")
  %>% drop_na() 
 ) -> RST_NoNA

 dim(RST_NoNA)
 head(RST_NoNA)


[1] 20 17

DAY_OF_MONTH DAY_OF_WEEK ORIGIN ORIGIN_STATE DEST DEST_STATE DEP_TIME
1  2           3           RST    MN           DEN  CO         1327    
2 12           6           RST    MN           ORD  IL          703    
3 13           7           RST    MN           DEN  CO         1205    
4 16           3           RST    MN           DEN  CO         1550    
5 16           3           RST    MN           ORD  IL          752    
6 17           4           RST    MN           MSP  MN          907    
  DEP_DELAY DEP_DEL15 ARR_TIME ARR_DELAY ARR_DEL15 CARRIER_DELAY WEATHER_DELAY
1 132       1         1415      89       1          89           0            
2  33       1          817      20       1          20           0            
3  50       1         1329      43       1          43           0            
4 275       1         1648     242       1         242           0            
5  82       1          906      69       1          69           0            
6 172       1          955     155       1         138           0            
  NAS_DELAY SECURITY_DELAY LATE_AIRCRAFT_DELAY
1 0         0               0                 
2 0         0               0                 
3 0         0               0                 
4 0         0               0                 
5 0         0               0                 
6 0         0              17

The following variation will exclude only NAs for the variable CARRIER_DELAY.

In [ ]:
#Getting the desired rows
(
  Ontime
  %>% filter(ORIGIN == "RST")
  %>% drop_na(CARRIER_DELAY	) 
 ) -> RST_NoNA_Carrier

 dim(RST_NoNA_Carrier)
 head(RST_NoNA_Carrier,25)

# Filter Action on Numbers


Consider again the RST dataframe for which there was a delay.

In [ ]:
 head(RST_NoNA_Carrier,25)

DAY_OF_MONTH DAY_OF_WEEK ORIGIN ORIGIN_STATE DEST DEST_STATE DEP_TIME
1   2           3           RST    MN           DEN  CO         1327    
2  12           6           RST    MN           ORD  IL          703    
3  13           7           RST    MN           DEN  CO         1205    
4  16           3           RST    MN           DEN  CO         1550    
5  16           3           RST    MN           ORD  IL          752    
6  17           4           RST    MN           MSP  MN          907    
7  17           4           RST    MN           MSP  MN         1840    
8  19           6           RST    MN           MSP  MN         1423    
9  23           3           RST    MN           ORD  IL         1326    
10 24           4           RST    MN           ORD  IL         1848    
11 24           4           RST    MN           ORD  IL          941    
12 24           4           RST    MN           MSP  MN          610    
13 24           4           RST    MN           MSP  MN         1235    
14 25           5           RST    MN           DEN  CO         1928    
15 28           1           RST    MN           MSP  MN          839    
16 28           1           RST    MN           MSP  MN         1855    
17 29           2           RST    MN           DEN  CO         1349    
18 29           2           RST    MN           MSP  MN         1830    
19 29           2           RST    MN           MSP  MN         1550    
20 30           3           RST    MN           MSP  MN         1243    
   DEP_DELAY DEP_DEL15 ARR_TIME ARR_DELAY ARR_DEL15 CARRIER_DELAY WEATHER_DELAY
1   132      1         1415       89      1          89              0         
2    33      1          817       20      1          20              0         
3    50      1         1329       43      1          43              0         
4   275      1         1648      242      1         242              0         
5    82      1          906       69      1          69              0         
6   172      1          955      155      1         138              0         
7    41      1         1932       30      1           0              0         
8   144      1         1508      136      1           0              0         
9  1200      1         1430     1175      1           0           1175         
10   82      1         1959       64      1           0             64         
11  191      1         1050      164      1           0            164         
12   -5      0          743       22      1           0              0         
13   36      1         1312       20      1           4              0         
14  483      1         2129      512      1         512              0         
15  144      1          933      132      1           0              0         
16   56      1         1939       36      1           0              0         
17  144      1         1536      159      1           0              0         
18   31      1         1938       35      1           0              9         
19    5      0         1711       31      1           0              0         
20   44      1         1354       62      1           0              0         
   NAS_DELAY SECURITY_DELAY LATE_AIRCRAFT_DELAY
1   0        0                0                
2   0        0                0                
3   0        0                0                
4   0        0                0                
5   0        0                0                
6   0        0               17                
7   0        0               30                
8   0        0              136                
9   0        0                0                
10  0        0                0                
11  0        0                0                
12 22        0                0                
13  0        0               16                
14  0        0                0                
15  0        0              132                
16  0        0        

The average carrier delay can be computed as follows.

$Average \space Carrier \space Delay = \frac{89+20+43+242+69+138+4+512}{8} = 139.625$

In [96]:
#Getting the average
(
  Ontime
  %>% filter(ORIGIN == "RST")
  %>% drop_na()
  %>% summarize(Avg_Carrier_Delay = mean(CARRIER_DELAY)) 
 ) 


Avg_Carrier_Delay
1 55.85

Question: Why do these average calcuations not agree?  

In [97]:
#Fixing the problem...
(
  Ontime
  %>% filter(ORIGIN == "RST")
  %>% drop_na()
  %>% filter(CARRIER_DELAY > 0)
  %>% summarize(Avg_Carrier_Delay = mean(CARRIER_DELAY)) 
 ) 


Avg_Carrier_Delay
1 139.625

Question:  What does the followign code do?  

In [ ]:
#Getting only RST
(
  Ontime
  %>% filter(ORIGIN == "RST")
  %>% drop_na()
  %>% filter(CARRIER_DELAY == 0)
 ) -> RST_NoNA_CarrierDelay_is_0

dim(RST_NoNA_CarrierDelay_is_0)
head(RST_NoNA_CarrierDelay_is_0,12)

[1] 12 17

DAY_OF_MONTH DAY_OF_WEEK ORIGIN ORIGIN_STATE DEST DEST_STATE DEP_TIME
1  17           4           RST    MN           MSP  MN         1840    
2  19           6           RST    MN           MSP  MN         1423    
3  23           3           RST    MN           ORD  IL         1326    
4  24           4           RST    MN           ORD  IL         1848    
5  24           4           RST    MN           ORD  IL          941    
6  24           4           RST    MN           MSP  MN          610    
7  28           1           RST    MN           MSP  MN          839    
8  28           1           RST    MN           MSP  MN         1855    
9  29           2           RST    MN           DEN  CO         1349    
10 29           2           RST    MN           MSP  MN         1830    
11 29           2           RST    MN           MSP  MN         1550    
12 30           3           RST    MN           MSP  MN         1243    
   DEP_DELAY DEP_DEL15 ARR_TIME ARR_DELAY ARR_DEL15 CARRIER_DELAY WEATHER_DELAY
1    41      1         1932       30      1         0                0         
2   144      1         1508      136      1         0                0         
3  1200      1         1430     1175      1         0             1175         
4    82      1         1959       64      1         0               64         
5   191      1         1050      164      1         0              164         
6    -5      0          743       22      1         0                0         
7   144      1          933      132      1         0                0         
8    56      1         1939       36      1         0                0         
9   144      1         1536      159      1         0                0         
10   31      1         1938       35      1         0                9         
11    5      0         1711       31      1         0                0         
12   44      1         1354       62      1         0                0         
   NAS_DELAY SECURITY_DELAY LATE_AIRCRAFT_DELAY
1   0        0               30                
2   0        0              136                
3   0        0                0                
4   0        0                0                
5   0        0                0                
6  22        0                0                
7   0        0              132                
8   0        0               36                
9   0        0              159                
10  4        0               22                
11 31        0                0                
12 18        0               44

Question:  What does the followign code do?  

In [ ]:
#Getting only RST
(
  Ontime
  %>% filter(ORIGIN == "RST")
  %>% drop_na()
  %>% filter(CARRIER_DELAY > 0, WEATHER_DELAY > 0)
 ) -> RST_NoNA_CarrierDelay_Above0_and_WeatherDelay_Above0

dim(RST_NoNA_CarrierDelay_Above0_and_WeatherDelay_Above0)
head(RST_NoNA_CarrierDelay_Above0_and_WeatherDelay_Above0,12)

[1]  0 17

DAY_OF_MONTH DAY_OF_WEEK ORIGIN ORIGIN_STATE DEST DEST_STATE DEP_TIME
     DEP_DELAY DEP_DEL15 ARR_TIME ARR_DELAY ARR_DEL15 CARRIER_DELAY
     WEATHER_DELAY NAS_DELAY SECURITY_DELAY LATE_AIRCRAFT_DELAY

Question:  What does the followign code do?  

In [ ]:
#Getting only RST
(
  Ontime
  %>% filter(ORIGIN == "RST")
  %>% drop_na()
  %>% filter(CARRIER_DELAY > 0 | WEATHER_DELAY > 0)
 ) -> RST_NoNA_CarrierDelay_Above0_or_WeatherDelay_Above0

dim(RST_NoNA_CarrierDelay_Above0_or_WeatherDelay_Above0)
head(RST_NoNA_CarrierDelay_Above0_or_WeatherDelay_Above0,12)

[1] 12 17

DAY_OF_MONTH DAY_OF_WEEK ORIGIN ORIGIN_STATE DEST DEST_STATE DEP_TIME
1   2           3           RST    MN           DEN  CO         1327    
2  12           6           RST    MN           ORD  IL          703    
3  13           7           RST    MN           DEN  CO         1205    
4  16           3           RST    MN           DEN  CO         1550    
5  16           3           RST    MN           ORD  IL          752    
6  17           4           RST    MN           MSP  MN          907    
7  23           3           RST    MN           ORD  IL         1326    
8  24           4           RST    MN           ORD  IL         1848    
9  24           4           RST    MN           ORD  IL          941    
10 24           4           RST    MN           MSP  MN         1235    
11 25           5           RST    MN           DEN  CO         1928    
12 29           2           RST    MN           MSP  MN         1830    
   DEP_DELAY DEP_DEL15 ARR_TIME ARR_DELAY ARR_DEL15 CARRIER_DELAY WEATHER_DELAY
1   132      1         1415       89      1          89              0         
2    33      1          817       20      1          20              0         
3    50      1         1329       43      1          43              0         
4   275      1         1648      242      1         242              0         
5    82      1          906       69      1          69              0         
6   172      1          955      155      1         138              0         
7  1200      1         1430     1175      1           0           1175         
8    82      1         1959       64      1           0             64         
9   191      1         1050      164      1           0            164         
10   36      1         1312       20      1           4              0         
11  483      1         2129      512      1         512              0         
12   31      1         1938       35      1           0              9         
   NAS_DELAY SECURITY_DELAY LATE_AIRCRAFT_DELAY
1  0         0               0                 
2  0         0               0                 
3  0         0               0                 
4  0         0               0                 
5  0         0               0                 
6  0         0              17                 
7  0         0               0                 
8  0         0               0                 
9  0         0               0                 
10 0         0              16                 
11 0         0               0                 
12 4         0              22